Worker and SharedWorker constructors should fail asynchronously in case of cross origin URLs

youennf opened the pull request:
#### e9a2fa13c71e43068c10d9394e78bb068a53f646
<pre>
Worker and SharedWorker constructors should fail asynchronously in case of cross origin URLs
<a href="https://rdar.apple.com/148229314">rdar://148229314</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290745">https://bugs.webkit.org/show_bug.cgi?id=290745</a>

Reviewed by Anne Van Kesteren.

Align with the spec and fail asynchronously in case of bad URL checks.
This aligns with the spec and Firefox behavior.
As discussed in <a href="https://github.com/web-platform-tests/wpt/issues/41745">https://github.com/web-platform-tests/wpt/issues/41745</a>, Chrome  also plans to align to the spec, and the web compatibility story seems favorable.
We add a runtime flag for now, in preview state, but off by default for now to further validate the web compatibility story.

We do a small refactoring to share more code between Worker and SharedWorker.
We update WPT tests from upstream to cope with the change of behavior and we update our local tests acccordingly.

* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/child-src/worker-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/child-src/worker-blocked.html:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/default-src/manifest-v3-default-src-block-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/worker-src/manifest-v3-worker-src-block-partial-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/worker-src/manifest-v3-worker-src-block-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/resources/child-src-test.js:
(window.onload):
(injectWorker):
* LayoutTests/http/tests/workers/resources/worker-redirect.js:
(testCrossOriginLoad.try.worker.onerror):
(testCrossOriginLoad):
* LayoutTests/http/tests/workers/worker-document-domain-security-expected.txt:
* LayoutTests/http/tests/workers/worker-document-domain-security.html:
* LayoutTests/http/tests/workers/worker-invalid-url-expected.txt:
* LayoutTests/http/tests/workers/worker-invalid-url.html:
* LayoutTests/http/tests/workers/worker-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/support/testharness-helper.js:
(async_test): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/fetch/metadata/generated/worker-dedicated-constructor.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/Worker_cross_origin_security_err-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/constructors/SharedWorker/same-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/constructors/Worker/same-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/dedicated-worker-in-data-url-context.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/dedicated-worker-import-failure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/shared-worker-import-failure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/shared-worker-in-data-url-context.window-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/fetch/metadata/generated/worker-dedicated-constructor.sub-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/workers/worker-redirect-expected.txt:
* LayoutTests/platform/wk2/http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-image-in-iframe-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/workers/AbstractWorker.cpp:
(WebCore::AbstractWorker::resolveURL):
(WebCore::AbstractWorker::validateURL):
* Source/WebCore/workers/AbstractWorker.h:
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::create):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::create):

Canonical link: <a href="https://commits.webkit.org/293014@main">https://commits.webkit.org/293014@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da5dd207ba20cc896fa4d8d1720da2f9f8e31904

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97711 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17336 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7553 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102818 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48240 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99756 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25789 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74423 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31612 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88328 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54772 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13131 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6244 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47683 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90387 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104819 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96334 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24791 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25163 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84488 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82905 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20879 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27475 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5154 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18393 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24752 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119959 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24574 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33666 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27888 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26148 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -9377,6 +9377,20 @@ WirelessPlaybackTargetAPIEnabled:
+  status: preview
+  category: dom
+  humanReadableName: "Storage API Estimate"
+  humanReadableDescription: "Enable Storage API Estimate"
```
This needs changing.

annevk commented on Source/WebCore/workers/AbstractWorker.cpp:
```diff
@@ -58,21 +58,27 @@ FetchOptions AbstractWorker::workerFetchOptions(const WorkerOptions& options, Fe
 
     // FIXME: This should use the dynamic global scope (bug #27887).
-    URL scriptURL = context.completeURL(url);
+    URL scriptURL = context->completeURL(url);
```
```suggestion
    auto scriptURL = context->completeURL(url);
```

annevk reviewed (approved):
Looks good modulo setting description.

youennf committed:
Worker and SharedWorker constructors should fail asynchronously in case of cross origin URLs
rdar://148229314
https://bugs.webkit.org/show_bug.cgi?id=290745

Reviewed by Anne Van Kesteren.

Align with the spec and fail asynchronously in case of bad URL checks.
This aligns with the spec and Firefox behavior.
As discussed in https://github.com/web-platform-tests/wpt/issues/41745, Chrome  also plans to align to the spec, and the web compatibility story seems favorable.
We add a runtime flag for now, in preview state, but off by default for now to further validate the web compatibility story.

We do a small refactoring to share more code between Worker and SharedWorker.
We update WPT tests from upstream to cope with the change of behavior and we update our local tests acccordingly.

* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/child-src/worker-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/child-src/worker-blocked.html:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/default-src/manifest-v3-default-src-block-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/worker-src/manifest-v3-worker-src-block-partial-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/extensions/manifest-v3/worker-src/manifest-v3-worker-src-block-wildcard-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/resources/child-src-test.js:
(window.onload):
(injectWorker):
* LayoutTests/http/tests/workers/resources/worker-redirect.js:
(testCrossOriginLoad.try.worker.onerror):
(testCrossOriginLoad):
* LayoutTests/http/tests/workers/worker-document-domain-security-expected.txt:
* LayoutTests/http/tests/workers/worker-document-domain-security.html:
* LayoutTests/http/tests/workers/worker-invalid-url-expected.txt:
* LayoutTests/http/tests/workers/worker-invalid-url.html:
* LayoutTests/http/tests/workers/worker-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/support/testharness-helper.js:
(async_test): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/fetch/metadata/generated/worker-dedicated-constructor.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/Worker_cross_origin_security_err-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/constructors/SharedWorker/same-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/constructors/Worker/same-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/dedicated-worker-in-data-url-context.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/dedicated-worker-import-failure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/modules/shared-worker-import-failure-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/shared-worker-in-data-url-context.window-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/fetch/metadata/generated/worker-dedicated-constructor.sub-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/workers/worker-redirect-expected.txt:
* LayoutTests/platform/wk2/http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-image-in-iframe-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/workers/AbstractWorker.cpp:
(WebCore::AbstractWorker::resolveURL):
(WebCore::AbstractWorker::validateURL):
* Source/WebCore/workers/AbstractWorker.h:
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::create):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::create):

Canonical link: https://commits.webkit.org/293014@main
